### PR TITLE
Add Sbus Non Inverted Mode

### LINKED ISSUE #390: Naze32 r.5 should support non/de/un-inverted s.bus configurations on Receiver Port Serial
It's fairly common the miniquad world to bypass the inverter on the x4r-sb radio via the a-holes mod.

This makes the radio send "normal" s.bus, i.e., un-inverted.

It'd be nice to support this.

### PR DESCRIPTION
**Pull request by jihlein**
Fixes #390.

This PR implements a non-inverted Sbus interface on the following targets: AQ32, Naze32, and Sparky.  The mode is called Sbus Non Inverted on the GCS setup screens.  The 3 targets implemented here give an F1, F3, and F4 example.

I've tested the capability to selected both Sbus and Sbus Non Inverted on all applicable ports on the above 3 targets.  Each target was able to boot and execute as expected.  As I don't have any Sbus receivers, I can not test further.  Hopefully @dustin and @Nub can test with SBus receivers.  Testing should include, at a minimum, both manual setup and wizard setup as applicable.  A quick test flight wouldn't hurt either.

As the port drop down boxes were getting long and somewhat disorganized, I alphabetized the selections so the lists don't look as disorganized.

A future PR should implement the Sbus Non Inverted mode on the remaining targets, along with the alphabetized selection list so all targets look uniform.

If there's a better way to implement some of this, let me know.

[Review on Reviewable](https://reviewable.io/reviews/d-ronin/dronin/403)

### CODE REVIEW

**Comment by fujin** [comment]
jenkins, build artifacts please

**Comment by tracernz** [comment]
This removes HoTT support on Naze?

**Comment by jihlein** [comment]
There was no HoTT support, so I cleaned up the drop down boxes.

**Comment by tracernz** [comment]
Looks like at least most of the code was there? Any reason not to complete it instead of removing?

**Comment by jihlein** [comment]
I think I goofed....

**Comment by jihlein** [comment]
HoTT support restored in Naze32.

**Comment by pug398** [comment]
Sparky1/X4R-SB with "A pad mod" seems fine. Wizard setup went smoothly and test flight had no issues. SBUS Non Inverted on Rx port.

**Comment by jihlein** [comment]
@mlyle, I think we should pull this from Renatus.  I figured out what you where talking about last night and am making the changes.  Reduced the argument count to PIOS_HAL by two.  That's the good news.  Bad news is it broke a few of targets (the ones not using PIOS_HAL yet), and it's going to take some time to fix those.  Brain, Colibri, and FlyingF3 are the culprits.  The rest went smoothly.  Should we write a PR to move those targets to PIOS_HAL and hold this one until the move is complete?

@pug398, glad to hear the concept works.

**Comment by jihlein** [comment]
Latest code pushed, not fully tested.  Brain, Colibri, and FlyingF3 do not compile.

**Comment by jihlein** [comment]
Minimal changes made to Brain, Colibri, and FlyingF3, everything compiles.

**Comment by pug398** [comment]
@jihlein Do you need a test on Brain?

**Comment by jihlein** [comment]
@pug398, Brain is not complete, it needs to move to PIOS_HAL before it can take full advantage of the SBus Non Inverted changes.  With the minimal changes I made, it should work as before.  PIOS_HAL for the Brain is going to take some thought.

**Comment by jihlein** [comment]
@pug38, can you rebuild with the latest files and try Sbus Non Inverted on Sparky1 again?

**Comment by pug398** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
artifacts here
http://jar.lyle.org:8080/job/dronin/61/Nodes=winx86/

Sorry I'm abusing jenkins at the moment

**Comment by pug398** [comment]
@jihlein quick setup and indoor flight. Everything still looks good. There is about a 10 second delay in the detection of stick movement on the max limit page but don't think that is related to this PR.

**Comment by mlyle** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
@pug398 -- did you test noninverted or inverted (or both?)

**Comment by pug398** [comment]
Non inverted and getting ready to test inverted.

**Comment by pug398** [comment]
Both doing it. It is late and I may be crazy but there is text about moving all controls and switches to max travel. At first the text takes up two lines but somewhere after a delay it blinks and text only takes up one line. At this point sticks will register. I also got a reset at this point a couple of times. Will check more tomorrow...

**Comment by mlyle** [comment]
Hm.  I did make a UI change to force you to get a minimal range on all channels before allowing you to advance past the max travel screen.  The "next" thing is greyed until all configured channels (including switches) have moved at least a minimum amount.  

That doesn't sound good-- especially the resets.  What TX/RX are you using?

I will try to get to testing input wizard again.

**Comment by mlyle** [comment]
Review status: 0 of 38 files reviewed at latest revision, 6 unresolved discussions.

---

*[flight/PiOS/Common/pios_hal.c, line 385 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YN-2WmnpOEw3KVoAa:-K7YN-2WmnpOEw3KVoAb:1548291931) ([raw file](https://github.com/d-ronin/dronin/blob/a688510a6ac6a9be88880bb644d897a18bc3e474/flight/PiOS/Common/pios_hal.c#L385)):*
This needs to not be a pointer, because as a pointer it often points to non-writable flash.  It should be struct pios_usart_cfg pios_hal_usart_temp_cfg;  Then the assignment can be done, I believe, with pios_hal_usart_temp_cfg = *usart_port_cfg.

---

*[flight/PiOS/Common/pios_hal.c, line 492 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YNk42S26fh_-VXDFo:-K7YNk42S26fh_-VXDFp:-1401440317) ([raw file](https://github.com/d-ronin/dronin/blob/a688510a6ac6a9be88880bb644d897a18bc3e474/flight/PiOS/Common/pios_hal.c#L492)):*
I think PIOS_USART_Init needs to be changed to not save a pointer to the config-- instead just holding the pointer to the regs.  Thankfully, the regs is already a pointer to a separate structure so it is not "copied" by the proposed assignment.

This is because pios_hal_usart_temp_cfg is on stack and only has scope/life as long until PIOS_HAL_ConfigurePort returns.

:/  Which is becoming a much more substantial change, because it breaks ChangeBaud as written, because it wants to see the init stuff later. 

---

*[flight/targets/naze32/board-info/board_hw_defs.c, line 724 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YQ9GbPfJxwwK_LvqS:-K7YQ9Gcm-Z8f5c0hbtJ:-583402638) ([raw file](https://github.com/d-ronin/dronin/blob/83469bb150cb8415010c85a368b6bee862f1e61c/flight/targets/naze32/board-info/board_hw_defs.c#L724)):*
I see you've de-constified this so it works.  Unfortunately this moves all this stuff to ram which really increases the pressure on F1 targets.

---

*[flight/targets/naze32/board-info/board_hw_defs.c, line 760 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YQ5ZqWsTlNBX_iD6v:-K7YQ5ZqWsTlNBX_iD6w:-1114765562) ([raw file](https://github.com/d-ronin/dronin/blob/83469bb150cb8415010c85a368b6bee862f1e61c/flight/targets/naze32/board-info/board_hw_defs.c#L760)):*
Probably shouldn't be renamed?

---

*[flight/targets/naze32/board-info/pios_board.h, line 157 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YQIzP5TVgOxjzBIM8:-K7YQIzP5TVgOxjzBIM9:812747157) ([raw file](https://github.com/d-ronin/dronin/blob/a688510a6ac6a9be88880bb644d897a18bc3e474/flight/targets/naze32/board-info/pios_board.h#L157)):*
Whitespace-- converted to space from tab.. should be consistent with below/above

---

*[flight/targets/pipxtreme/fw/pios_board.c, line 164 \[r5\]](https://reviewable.io:443/reviews/d-ronin/dronin/403#-K7YQSIh5Tftf8QnJ9Vh:-K7YQSIh5Tftf8QnJ9Vi:-1787918134) ([raw file](https://github.com/d-ronin/dronin/blob/a688510a6ac6a9be88880bb644d897a18bc3e474/flight/targets/pipxtreme/fw/pios_board.c#L164)):*
space at end of line

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/403)*

**Comment by mlyle** [comment]
I'm sorry this is proving a pain.  i didn't fully understand what you did when I started reviewing.  This is much closer, but I'd rather not move the configuration structures to RAM if it can be avoided.  Unfortunately, that takes some degree of refactoring of the pios_usart implementations.

**Comment by mlyle** [comment]
No choice but to defer this, with other churn and concern about sbus maturity in general.  Will send a PR to this in a few days fixing the usart half.

**Comment by jihlein** [comment]
After ignoring this for a day or so, I believe I have the usart refactoring figured out after taking another look at it.  Will test my changes over the next day or so and let you know.

**Comment by jihlein** [comment]
Note: Waiting for #411, #413, #414, and #444 to merge.

**Comment by jihlein** [comment]
All targets but Brain now compiling.

**Comment by jihlein** [comment]
Removed WIP from title.  All targets building.  Ready for testing and review.

Let the fun begin.....

**Comment by mlyle** [comment]
Not going to mess with this for a few days until I feel a little better about next, though.

File: flight/PiOS/Common/pios_hal.c [hunk 1]
```diff
@@ -379,17 +382,144 @@ void PIOS_HAL_ConfigurePort(HwSharedPortTypesOptions port_type,
...
+		PIOS_HAL_ConfigureCom(usart_port_cfg, &usart_port_params, 0, PIOS_COM_DEBUGCONSOLE_TX_BUF_LEN, com_driver, &port_driver_id);
+		target = &pios_com_debug_id;
+#endif  /* PIOS_INCLUDE_DEBUG_CONSOLE */
+		break;
+
+	case HWSHARED_PORTTYPES_DSM:
+#if defined(PIOS_INCLUDE_DSM)
+		if (dsm_cfg && usart_port_cfg) {
+			usart_port_params.init.USART_BaudRate              = 115200;
+			//usart_port_params.init.USART_WordLength          = USART_WordLength_8b;
```
Comment (mlyle) [inline flight/PiOS/Common/pios_hal.c]: Remove commented code

File: ground/gcs/src/plugins/coreplugin/iboardtype.h [hunk 2]
```diff
@@ -163,6 +163,7 @@ class CORE_EXPORT IBoardType : public QObject
         INPUT_TYPE_PPM,
         INPUT_TYPE_DSM,
         INPUT_TYPE_SBUS,
+		INPUT_TYPE_SBUSNONINVERTED,
```
Comment (mlyle) [inline ground/gcs/src/plugins/coreplugin/iboardtype.h]: Indentation (need to match the wrong indentation on other lines ):

**Comment by mlyle** [comment]
Looks very nice.  I owe you a more thorough read.  net savings of 2700 lines is very nice :)

**Comment by jihlein** [comment]
Regarding the commented code in PIOS_HAL.c, no problem removing it but it was intentionally left as documentation to what the complete port setting would be.  That would be the only reason to consider leaving it in place.

**Comment by mlyle** [comment]
Alternatively, uncomment; optimizer should take it out.

**Comment by jihlein** [comment]
If that's the case, I'd prefer to uncomment it, leaving it there for documentation's sake.

File: ground/gcs/src/plugins/coreplugin/iboardtype.h (hunk 2 above)
Comment (tracernz) [inline ground/gcs/src/plugins/coreplugin/iboardtype.h]: We use spaces in GCS code.

**Comment by mlyle** [comment]
Also needs a rebase unfortunately; probably the revo changes.

**Comment by jihlein** [comment]
I loaded this on a Sparky, did a quick test of external mag setup on the flex port, port configured correctly.  

I'll try to get this loaded on my AQ32 tricopter in the near future.  That will cover gps config, msp config, openlog config, external mag config, FrSky sensor hub config, and ppm config.

Is there someone who could run a test of sbus non inverted?  Any target will do.

**Comment by dustin** [comment]
I could do something on a desk.  I've got radios that do SBNI

**Comment by mlyle** [comment]
I will try to get to it soon.  I was hoping to merge it to my "kitchensink" branch, but there's some conflicts with the things I'm doing with pipxtreme so it should probably wait to the next round of test (but possibly merge first).

**Comment by dustin** [comment]
Looking to do some testing here.  I started by attempting to rebase it over next.  It's got a bunch of trailing whitespace, which rebase complained about, but I'm not entirely sure if that caused the issues, but it's a thing to resolve.

(testing it as-is)

---

Review status: 0 of 52 files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/403#-:-K9opXf7L27ZHcYnORC_:-1648984712)_

**Comment by jihlein** [comment]
I just finished rebasing it.  Did a bunch of testing with AQ32 today.  Here's what I verified:

1)PPM rx setup
2)GPS operation
3)External mag operation and orientation
4)FrSky sensor hub operation
5)MSP operation (note a bug with stack sizes was found here and corrected on next)
6)VCP pass thru with a baud rate change to 115200 from 57600 (used to setup MWOSD)
7)OpenLog operation

All ports setup and functioned as expected.  Tests of Sbus and Sbus non inverted would be helpful, as would testing on an F1 and F3 target.  I've done some basic tests on Sparky.

I've got a little more setup work to do, but hope to fly the AQ32 target tomorrow.

**Comment by mlyle** [comment]
```
/var/lib/jenkins/jobs/dron/workspace/Nodes/master/build/fw_flyingf3/logging.o: In function `LoggingInitialize':
/var/lib/jenkins/jobs/dron/workspace/Nodes/master/flight/Modules/Logging/logging.c:149: undefined reference to `pios_com_openlog_logging_id'
/var/lib/jenkins/jobs/dron/workspace/Nodes/master/build/fw_flyingf3/pios_hal.o: In function `PIOS_HAL_ConfigurePort':
/var/lib/jenkins/jobs/dron/workspace/Nodes/master/flight/PiOS/Common/pios_hal.c:655: undefined reference to `pios_com_openlog_logging_id'
collect2: error: ld returned 1 exit status
```

**Comment by mlyle** [comment]
Looks like a really clean/nice history btw.

**Comment by dustin** [comment]
I couldn't get my x6r working on it with a sparky 1.  It worked OK on next.

I also failed to get my 4ch remix working in either direction.  It didn't work on next either, though, so I might just be doing it wrong.  heh.

---

Review status: 0 of 52 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/403#-:-K9p6oYKXVb9Lg_jMh4O:1391826377)_

**Comment by pug398** [comment]
jenkins, build artifacts please

**Comment by mlyle** [comment]
@jihlein it looks like there's something wrong with this rebase-- did you base off upstream/next?   There's alreayd-commited PRs in the list.

@pug398 -- the build failed, it looks like.  Also I'm not sure if this is worth testing yet, as it wasn't working last night.

**Comment by jihlein** [comment]
Rebased to current next.

Flew the AQ32 setup on a tricopter as described a few comments above with no issues, all ports setting up correctly and flew with no surprises.  Pulled telemetry off OpenLog and parse cleanly with Octave.

Setup a Sparky with DSM satellite on RvcrPort, also on a tricopter.  Everything setup correctly and flew with no issues.

**Comment by pug398** [comment]
@jihlein I have SB/SBNI rx mounted on Brain quad if you still need another target tested.

**Comment by jihlein** [comment]
@pug398, that would be perfect, the sb/sbni still need some testing/verification.  On the Brain FC, even better.

**Comment by mlyle** [comment]
I believe @dustin tried this and I don't think it worked (either non-inverted or inverted).  I have desk-checked it and didn't see anything obviously wrong, but have not gotten a chance to try it myself.

**Comment by mlyle** [comment]
Man, I read the code again and it sure looks like it should work.  If it's broken, it's something subtle and dumb :P

**Comment by jihlein** [comment]
I believe it was tried on Sparky, and there was a problem that was corrected with the last commit.

**Comment by pug398** [comment]
jenkins, build artifacts please

**Comment by pug398** [comment]
Ok so this did not pull in old settings and brain needed to be fully erased. With 90 changes I guess that can be expected :) With a settings import it flew fine on sbus and moving rx to sbni output it flew fine with just a main port change to sbus non inverted. A full erase and setup using radio wizard needs to be done as that is where issues usually show up. Will try to get to that tomorrow afternoon. Looking good so far.

**Comment by jihlein** [comment]
Very cool @pug398, that's great to hear. Thanks!

**Comment by tracernz** [comment]
@mlyle Jenkins didn't call back with the artifacts but it has finished building them. Does it not do that with merge conflicts?

**Comment by mlyle** [comment]
It's supposed to and the comment script run, but it apparently didn't happen.  It just uses the pull id, which was set right in the build.  So maybe there was a github glitch or something (I don't check success status or retry failures).

**Comment by pug398** [comment]
On a fresh Brain setup using sbus non inverted, input setup via radio wizard completed without any problems. Subsequent flights went well.

**Comment by jihlein** [comment]
Not sure why, but the make ef_target cmds are are failing.

recipe for target 'c:/dronin/build/ef_brain/ef_brain.fw.pre.pad' failed, for example.  AQ32 and Sparky fail the same way.

**Comment by tracernz** [comment]
@jihlein Can't replicate that on Linux. Can you open an issue with further details?

**Comment by jihlein** [comment]
Rebased to current next earlier today.  Here's the current test matrix status.

image

X1: AQ32 test of combridge also changed the default baud rate of 57600 to 115200 to talk to and setup an MWOSD, effectively checking the changes to PIOS_USART_ChangeBaud on the F4 platform.

X2: Sparky test of  telemetry also changed the default baud rate of 57600 to 115200 to talk to bluetooth radio, effectively checking the changes of PIOS_USART_ChangeBaud on the F3 platform.

**Comment by mlyle** [comment]
Would be nice to see sbus/sbusni work on F3.  Also FrskySport is different on F3 too.

**Comment by pug398** [comment]
sbus/sbusni flight tested on sparky1 no problems. FrskySport tested on sparky1 main port with basic telemetry displayed on Taranis. No problems indicated.

**Comment by pug398** [comment]
Bluetooth telemetry tested ok on sparky flexi port @115k as well.

**Comment by mlyle** [comment]
I have skimmed this and it looks good to me-- I'm getting a little blind to it from reading it several times.

It has done well on all the key tests.

If there's no objections I'm inclined to merge this and shake it out the rest of the way on next.

**Comment by tracernz** [comment]
Reviewed 1 of 20 files at r3, 1 of 4 files at r5, 18 of 50 files at r6, 6 of 16 files at r7, 18 of 25 files at r8, 6 of 6 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/d-ronin/dronin/403)_

**Comment by jihlein** [comment]
Updated test matrix.

**Comment by tracernz** [comment]
S.BUS on F1 would be good if anyone has some hardware available to test.

**Comment by mlyle** [comment]
I don't expect F1 will be different than F4-- it works the same all in all, doesn't it?

We're accumulating a fair bit of test debt and minor regression, so it'll be time to shake things out thoroughly soon.  Best if it happens together IMO.

**Pull request merged**